### PR TITLE
SQLite: Client-eval more expressions

### LINKED ISSUE #10534: SQLite Query: Should we be translating decimal/TimeSpan operations?
Any operation (besides equality comparison) on `decimal` values will be lossy on SQLite since they are implicitly converted to `REAL` on the server. Should we continue translating them?

### PR DESCRIPTION
**Pull request by bricelam**
Many operations involving the following types are problematic and should not be performed on the server:
- DateTime
- DateTimeOffset
- decimal
- TimeSpan
- ulong

Fixes #10534

### CODE REVIEW

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteResultOperatorHandler.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+using Remotion.Linq.Clauses;
+using Remotion.Linq.Clauses.ResultOperators;
+
+namespace Microsoft.EntityFrameworkCore.Sqlite.Query.Internal
+{
+    /// <summary>
+    ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+    ///     directly from your code. This API may change or be removed in future releases.
+    /// </summary>
+    public class SqliteResultOperatorHandler : RelationalResultOperatorHandler
```
Comment (bricelam) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteResultOperatorHandler.cs]: ☹️ `.Internal` API

Comment (smitpatel) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteResultOperatorHandler.cs]: Pains of provider writers :trollface:

File: src/EFCore.Relational/Query/Internal/RelationalResultOperatorHandler.cs [hunk 2]
```diff
@@ -1007,7 +1007,11 @@ var sqlTranslatingExpressionVisitor
                     typeof(bool)));
         }
 
-        private static void PrepareSelectExpressionForAggregate(SelectExpression selectExpression)
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        protected static void PrepareSelectExpressionForAggregate(SelectExpression selectExpression)
```
Comment (anpete) [inline src/EFCore.Relational/Query/Internal/RelationalResultOperatorHandler.cs]: virtual instance more useful?

File: src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -32,54 +33,70 @@ public class SqliteSqlTranslatingExpressionVisitor : SqlTranslatingExpressionVis
         {
         }
 
+        private static readonly IReadOnlyCollection<Type> _negationRestrictedTypes
+            = new HashSet<Type> { typeof(decimal), typeof(TimeSpan) };
```
Comment (anpete) [inline src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Consider just hard-coding the types in the condition below, this set may be overkill

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteResultOperatorHandler.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            var selectExpression = relationalQueryModelVisitor.TryGetQuery(queryModel.MainFromClause);
+
+            if (!relationalQueryModelVisitor.RequiresClientResultOperator
+                && selectExpression != null
+                && selectExpression.Projection.Count == 1
+                && _restrictedOperators.TryGetValue(resultOperator.GetType(), out var restrictedTypes))
+            {
+                PrepareSelectExpressionForAggregate(selectExpression);
+
+                if (!(selectExpression.Projection.First().RemoveConvert() is SelectExpression))
```
Comment (anpete) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteResultOperatorHandler.cs]: Projection[0] - alloc

**Comment by anpete** [comment]
Approach looks good

File: src/EFCore.Relational/Query/Internal/RelationalResultOperatorHandler.cs (hunk 2 above)
Comment (bricelam) [reply src/EFCore.Relational/Query/Internal/RelationalResultOperatorHandler.cs]: I think the whole class could be more overridable (everything is `private static`), but it's an .Internal class so the priority seemed low.

**Comment by bricelam** [comment]
OK, this change is done and ready for review.

lol, after writing all those tests, I'm shocked nobody has reported any of these issues (besides `decimal` comparisons) @ajcvickers and @divega Is this something we want to consider patching in 2.1?

**Comment by adrianhara** [comment]
Starting which version does this behavior apply? We're using 2.2.1 and (from the logging at least) it looks like decimal comparison (order by) is done in the sql/db, not client-side.

**Comment by bricelam** [comment]
@adrianhara Order-by didn't make it into this PR. Issue #11925 is tracking that work. You can force client eval like this:
``` csharp
myQuery.AsEnumerable().OrderBy(e => e.MyDecimalProperty);
```

**Comment by adrianhara** [comment]
@bricelam thanks. My use case is i DON'T want client eval for order by for a decimal, because we store it as a padded string (e.g. "0000012.34000") and orderby works server-side.

**Comment by bricelam** [comment]
I think we’re improving this in 3.0. If you use a value converter to convert it to/from TEXT it should server-eval. You can probably do it in 2.x by using a string property on your entity and an unmapped decimal property that converts it.

**Pull request merged**